### PR TITLE
feat: Replace Picasso Image Library with Glide

### LINKED ISSUE #462: Replace Picasso with Glide
Even after using the latest picasso version, it is creating a lot of problems and bugs related to image loading. Sometimes images are not loaded at first time and sometimes even the ones loaded previously are not

@mejariamol Can you please take this up?

### PR DESCRIPTION
**Pull request by mejariamol**
fixes #462

### CODE REVIEW

**Comment by mejariamol** [comment]
@iamareebjamal please review.

File: app/proguard-rules.pro [hunk 1]
```diff
@@ -67,3 +67,5 @@
 -keep public class com.evernote.android.job.v14.PlatformAlarmReceiver
 -keep public class com.evernote.android.job.JobBootReceiver
 -keep public class com.evernote.android.job.JobRescheduleService
+
+-keep public class * implements com.bumptech.glide.module.GlideModule
```
Comment (iamareebjamal) [inline app/proguard-rules.pro]:
```
-keep public class * implements com.bumptech.glide.module.GlideModule
-keep public class * extends com.bumptech.glide.AppGlideModule
-keep public enum com.bumptech.glide.load.resource.bitmap.ImageHeaderParser$** {
  **[] $VALUES;
  public *;
}

# for DexGuard only
-keepresourcexmlelements manifest/application/meta-data@value=GlideModule
```

File: app/build.gradle [hunk 2]
```diff
@@ -101,6 +101,9 @@ dependencies {
     compile "com.android.support:cardview-v7:${support_lib_version}"
     compile "com.android.support:recyclerview-v7:${support_lib_version}"
     compile "com.android.support:design:${support_lib_version}"
+    compile "com.github.bumptech.glide:glide:4.0.0"
+    compile "com.github.bumptech.glide:okhttp3-integration:4.0.0"
+    annotationProcessor "com.github.bumptech.glide:compiler:4.0.0"
```
Comment (iamareebjamal) [inline app/build.gradle]: Please move annotationProcessor to bottom

**Comment by mejariamol** [comment]
>Can't see GlideRequest in commit. Is it auto generated?

Yeah.

File: app/proguard-rules.pro [hunk 3]
```diff
@@ -67,3 +67,6 @@
 -keep public class com.evernote.android.job.v14.PlatformAlarmReceiver
 -keep public class com.evernote.android.job.JobBootReceiver
 -keep public class com.evernote.android.job.JobRescheduleService
+
+-keep public class * extends com.bumptech.glide.module.AppGlideModule
+-keep class com.bumptech.glide.GeneratedAppGlideModuleImpl
```
Comment (iamareebjamal) [inline app/proguard-rules.pro]: This should be 

```
-keep public class * implements com.bumptech.glide.module.GlideModule
-keep public class * extends com.bumptech.glide.AppGlideModule
-keep public enum com.bumptech.glide.load.resource.bitmap.ImageHeaderParser$** {
  **[] $VALUES;
  public *;
}
```

**Comment by iamareebjamal** [comment]
@mejariamol Please follow https://github.com/bumptech/glide/wiki/Configuration#keeping-a-glidemodule

Tag is to be added in AndroidManifest.xml 

Also, disk cache should be enabled and set to 10 MB

File: app/build.gradle [hunk 4]
```diff
@@ -112,7 +114,6 @@ dependencies {
     compile "com.github.raizlabs.dbflow:dbflow-core:${dbflow_version}"
     compile "com.github.raizlabs.dbflow:dbflow:${dbflow_version}"
     compile "com.github.raizlabs.dbflow:dbflow-rx2:${dbflow_version}"
-    compile 'com.squareup.picasso:picasso:3.0.0-SNAPSHOT'
```
Comment (iamareebjamal) [inline app/build.gradle]: Also remove the SONATYPE repo configuration above

**Comment by mejariamol** [comment]
>Tag is to be added in AndroidManifest.xml

That's not required in version 4

About disk cache, I have kept it in `TODO`. But will do it in now.

**Comment by iamareebjamal** [comment]
Also, the image loading is not working. It is using the host selection interceptor and replacing the hosts from all url to open event server. Please fix it too

**Comment by iamareebjamal** [comment]
Circle Image is not working as well

**Comment by mejariamol** [comment]
>Also, the image loading is not working. It is using the host selection interceptor and replacing the hosts from all url to open event server. Please fix it too

which image. in the nav header? It is working for me.

**Comment by iamareebjamal** [comment]
It was not working on first load. Perhaps use 2 OkHttp clients. One base without host interceptor and one with it for Retrofit

**Comment by mejariamol** [comment]
@iamareebjamal please give me your credentials you tested imge loading on.

**Comment by iamareebjamal** [comment]
It's not about the credentials. I modified the local db and tested it. I can send you the logs. I removed the okhttp client and it worked fine. For now, let's just remove that from glide as it already logs the failed requests

Please add this too http://bumptech.github.io/glide/int/recyclerview.html

**Comment by mejariamol** [comment]
doing it.

**Comment by mejariamol** [comment]
@iamareebjamal please test now.

File: app/proguard-rules.pro (hunk 3 above)
Comment (mejariamol) [inline app/proguard-rules.pro]: I have followed v4 doc. This is changed in v4 I guess.

**Comment by iamareebjamal** [comment]
@mejariamol circular image is still not working

**Comment by mejariamol** [comment]
>circular image is still not working

which one?

File: app/src/main/java/org/fossasia/openevent/app/common/app/glide/GlideAPI.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import com.bumptech.glide.module.AppGlideModule;
+
+import java.io.InputStream;
+
+@GlideModule
+public final class GlideAPI extends AppGlideModule {
+
+    @Override
+    public void registerComponents(Context context, Glide glide, Registry registry) {
+        registry.replace(GlideUrl.class, InputStream.class, new OkHttpUrlLoader.Factory());
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/app/glide/GlideAPI.java]: This isn't needed if we are not using logger. Please remove this and add a comment to use logger in the future

Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/common/app/glide/GlideAPI.java]: then it will use HttpConnection. right?

**Comment by iamareebjamal** [comment]
Please refactor binding adapters like this:

```Java
private static void setGlideImage(ImageView imageView, String url, Drawable drawable, Transformation<Bitmap> transformation) {
        if (TextUtils.isEmpty(url)) {
            if (drawable != null)
                imageView.setImageDrawable(drawable);
            return;
        }

        GlideRequest<Drawable> request = GlideApp
            .with(imageView.getContext())
            .load(Uri.parse(url));

        if (drawable != null) {
            request
                .placeholder(drawable)
                .error(drawable);
        }

        request
            .centerCrop()
            .transition(withCrossFade())
            .transform(transformation != null ? transformation : new CenterCrop())
            .into(imageView);
    }
```

```Java
@BindingAdapter(value = {"circleImageUrl", "placeholder"}, requireAll = false)
    public static void bindCircularImage(ImageView imageView, String url, Drawable drawable) {
        setGlideImage(imageView, url, drawable, new CircleCrop());
    }
```

**Comment by iamareebjamal** [comment]
> which one?
@mejariamol Event thumbnail and user avatar

**Comment by mejariamol** [comment]
@iamareebjamal please test now.

**Comment by mejariamol** [comment]
@iamareebjamal build is fixed. please review.

**Pull request merged**